doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,266 @@
+
+```c
+    p = OSSL_PARAM_locate_const(params, OSSL_ALG_PARAM_STRICT_CHECKS);
+    if (p != NULL
+        && !OSSL_PARAM_get_int(p, &ctx->strict_checks))
+        return 0;
+```
+
+The setter is initially -1 (unknown) and can be set to 0 or 1 via a set_ctx call.
+If the setter is -1, when the FIPS related approved mode check is done then it
+uses the value from the FIPS configuration instead.
+
+- To access the indicator via code requires a getter (e.g OSSL_ALG_PARAM_APPROVED_INDICATOR),
+
+```c
+    p = OSSL_PARAM_locate(params, OSSL_ALG_PARAM_APPROVED_INDICATOR);
+    if (p != NULL && !OSSL_PARAM_set_int(p, ctx->approved))
+        return 0;
+```
+
+This initially has a value of -1, and is set to either 0 or 1 when a FIPS approved
+mode check is done. The getter allows you to access the indicator value after the
+operation has completed.
+
+If strict_checks is set to 1 then:
+- the FIPS configuration value is not used
+- If the operation is detected to be using unapproved mode then an error will
+  occur and the indicator callback will not be triggered.
+
+- Example Algorithm Check
+
+```c
+void alg_init(ALG_CTX *ctx)
+{
+    ctx->strict_checks = -1;
+    ctx->approved = -1;
+}
+
+int alg_check_approved(ALG_CTX *ctx)
+{
+    int pass;
+
+    ctx->approved = 1;
+    pass = some_fips_test_passes(ctx->libctx); // Check FIPS restriction for alg
+    if (!pass) {
+        ctx->approved = 0;
+        if (ctx->strict_checks == -1)
+            ctx->strict_checks = fips_config_get(ctx->libctx, op);
+        if (ctx->strict_checks != 0
+                || !OSSL_INDICATOR_callback(ctx->libctx, "ALG NAME", "ALG DESC"))
+        return 0;
+    }
+    return 1;
+}
+```
+
+- Existing security check changes
+
+OpenSSL already uses FIPS configuration options to perform security_checks, but
+the existing code needs to change to work with indicators.
+
+e.g. existing code
+
+```c
+    if (ossl_securitycheck_enabled(ctx)) {
+       pass = do_some_alg_test(ctx);
+       if (!pass)
+        return 0; /* Produce an error */
+    
+    }
+```
+In updated code for indicators the test always runs.. i.e.
+
+```c
+    pass = do_some_alg_test(ctx);
+    // Do code similar to alg_check_approved() above
+    // which will conditionally decide whether to return an error
+    // or trigger the indicator callback.
+```
+
+Issues with setting OSSL_ALG_PARAM_STRICT_CHECKS
+------------------------------------------------
+
+Normally a user would set params such as OSSL_ALG_PARAM_STRICT_CHECKS using
+set_ctx_params() but some algorithms currently do checks in their init operation.
+These init functions normally pass an OSSL_PARAM[] argument, but this still
+requires the user to set OSSL_ALG_PARAM_STRICT_CHECKS in their init.
+
+e.g.
+
+```c
+int strict = 0;
+params[0] = OSSL_PARAM_construct_int(OSSL_ALG_PARAM_STRICT_CHECKS, strict);
+EVP_DigestSignInit_ex(ctx, &pctx, name, libctx, NULL, pkey, params);
```

Comment:
Suggest: `params[1] = OSSL_PARAM_construct_end();`
so the code doesn't walk memory when copy/pasted.

Reply:
This is for illustration purposes only. I do not declare params[2] = { END, END };

Reply:
It will, nonetheless, be blindly copied...
The documentation should be exemplary.